EIPS/eip-6963.md

```diff
@@ -46,6 +46,7 @@ Provider Discovery Library: A library or piece of software that assists a DApp t
 Each wallet provider will be announced with the following interface `EIP6963ProviderInfo`. The values in the `EIP6963ProviderInfo` MUST be used as follows:
 
 - **`uuid`** - a globally unique of the wallet provider that MUST be (UUIDv4[^rfc4122] compliant) to uniquely distinguish different [EIP-1193](./eip-1193.md) provider sessions. The cryptographic uniqueness provided by UUIDv4[^rfc4122] guarantees that two independent `EIP6963ProviderInfo` objects can be separately identified.
+- **`rdns`** - an optional identifier that is used to optimistically identify the wallet by the Decentralized Application. If supplied by the wallet it MUST utilize an reverse DNS notation identifier that matches the domain name of the wallet provider.
```

Comment:
> _to optimistically identify the kind of wallet being injected_

I'm totally confused by the concept here. Can you please explain what info can be _optimistically_ obtained from an RDN property?

The announced wallet provider already contains _name_, _icon_ and an Ethereum provider object, which is everything that is needed for a dapp to connect to this wallet and interact with it.
More so, this data is available immediately as the event is received by the dapp. I do not understand what "optimistic" mechanism is being made available here.

Reply:
They are needed for “connect wallet” libraries like Web3Modal, RainbowKit and ConnectKit to achieve a robust and non-buggy experience, something that has been very hard to manage with flags on the EIP-1193 provider.

https://github.com/ethereum/EIPs/pull/7167/files#diff-927530d6d340560a972fe4fc1b91d7833e7a959089cc9c9d2baee54dd128c410R237

Reply:
'Optimistically' wasn't my word, but I interpreted it as a reference to the leap of faith that the wallet features are [still] the features the dapp thinks that vendor provides? In the sense that the legacy, unverifiable `isMetamask` flags can only be trusted by an optimist on a happy path😅.

Would it still be an "optimistic" identification process if the DNS string were a valid one under the control of the wallet vendor? In the telegram group there was some speculation that this could be an extension point for verifiable identification, publishing proof-of-control of, say, a package manager or app-store identifier at that address...

Reply:
> In the sense that the legacy, unverifiable isMetamask flags can only be trusted by an optimist on a happy path

Yes, this is the intention behind that word when I added it in there.

The `providerId` and `rdns` both act as identifiers for previously interacted wallets, but only in the case where wallets behave as expected. In the following events the expected behavior is undefined:

1. When the user deletes their wallet and switches so the wallet is no longer available
2. When the wallet changes the identifier used
3. When the wallet doesn't supply a compliant identifier or the identifier is unknown
4. When a wallet overlaps it's identifier (either intentionally or not) with another

In all of these cases, the `providerId` and `rdns` properties will potentially behave in undefined ways according to the spec. These are just the ones that I've come up and there's likely to be more implementation specific ones that arise overtime. Essentially we're leaving this part of the UX up to the DApp connection library implementers (or independent implementers) so if the implementers opt to not handle one of these cases or new implementation specific ones then the UX will behave in undefined ways and produce bugs.

Reply:
I guess my issue is with term itself. "Optimistic" conventionally means a situation where a program assumes some state which will be verified or discarded later (usually this means a response from the server which is considered to be source of truth)

In our case, I don't see any of such mechanics.
* If we're talking about feature detection based on rdns string, then this is a terrible idea (see https://webaim.org/blog/user-agent-string-history/) (but I'm not sure we're talking about that)
* If the dapp is going to _verify_ the rdns value by making an async rpc request to the wallet, then the term "optimistic" does make sense. But I've never heard or such rpc api.

Reply:
> If we're talking about feature detection based on rdns string, then this is a terrible idea (see [webaim.org/blog/user-agent-string-history](https://webaim.org/blog/user-agent-string-history/)) (but I'm not sure we're talking about that)

Great point here! We are litterally reinventing the UA string and will almost certainly re-encounter all it's problems along the way. My understanding is this is the primary reason that others have been wanting this and this example really exemplifies why this is a bad idea.

Reply:
Maybe we could find a better replacement of rdns, but I think making it optional would remove the value it's meant to have

Reply:
Well, I think the `isMetamask` flag is a better analogy to UA string, in that it's self-attested and people are highly incentivized to use it. thus my idea of making an optional verification step when it mattered that relied on DNS control...

Reply:
We also should consider that this new EIP will offer a new environment when isMetaMask flag won't be as useful as before, why would a wallet impersonate another one when it's going to be recognized by the dapp and be shown as wanted?

Reply:
well, today there are good and bad reasons why a non-metamask wallet might set that flag:
- metamask fork or clone has all the same interfaces and UX flows, makes it easy for dapps to integrate by announcing equivalent features
- phishing wallet presents user and dapp alike metamask interface but does something else under the hood (usually bad)

i'm presuming there would be both good and bad use-cases might still apply for `rdns`, but in the latter case some kind of verification extension would prevent it, and in the former case, (which I consider a feature and not a bug for wallet competition, evolution, and adoption), subdomains could work wonders here, as in:
`xyz.upandcoming.compatibilityMode.metamask` . this way, a dapp that already implemented a Metamask-style flow and doesn't want to support upandcoming.xyz's bespoke flow could just match on `compatibilityMode.metamask` and assume the metamask integration will work for that wallet 😎

Reply:
> why would a wallet impersonate another one when it's going to be recognized by the dapp and be shown as wanted

That is highly dependent on whether the web page will choose to discriminate based on wallets. One reasoning could be because the wallet doesn't support an RPC call they need as an example and get's published as an IPFS Static DApp so rather than update the DApp, future wallets that add support for the RPC call may be forced to imitate others.